### PR TITLE
Muon - Core Code; Load Widget Utilities

### PR DESCRIPTION
**Pull request by TWJubb**
Refs #23642 

### Summary

This PR creates a set of utility functions (with unit tests), for use with the load widget.

-----------

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by AnthonyLim23** [comment]
I think we should create a utils and a threading folder.

File: scripts/Muon/GUI/Common/muon_file_utils.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
+from __future__ import (absolute_import, division, print_function)
+
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/muon_file_utils.py]: copyright is missing

File: scripts/Muon/GUI/Common/muon_file_utils.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    If instrument is supported, returns the directory name on the ISIS network
+    in which its data can be found
+    """
+    if instrument in allowed_instruments:
+        instrument_directory = instrument
+        if instrument == "CHRONUS":
+            instrument_directory = "NDW1030"
+        return instrument_directory
+    else:
+        return None
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/muon_file_utils.py]: I think we should have some sort of warning here

File: scripts/Muon/GUI/Common/muon_file_utils.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    """
+    If instrument is supported, attempts to find the file on the ISIS network which
+    contains the data from its current (most up-to-date) run.
+    """
+
+    instrument_directory = get_instrument_directory(instrument)
+    if instrument_directory is None:
+        return ""
+
+    autosave_file_name = _instrument_data_directory(instrument) + FILE_SEP + "autosave.run"
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/muon_file_utils.py]: looking at the `_instrument_data_directory` method it calls `instrument_directory`, which is above. Why can we not pass in the instrument directory?

File: scripts/Muon/GUI/Common/muon_file_utils.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    autosave_points_to = ""
+    if not check_file_exists(autosave_file_name):
+        raise ValueError("Cannot find file : " + autosave_file_name)
+    with open(autosave_file_name, 'r') as autosave_file:
+        for line in autosave_file:
+            if os.path.isfile(line):
+                autosave_points_to = line
+    if autosave_points_to == "":
+        # Default to auto_A (replicates MuonAnalysis 1.0 behaviour)
+        current_run_filename = FILE_SEP + FILE_SEP + instrument_directory + FILE_SEP + "data" \
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/muon_file_utils.py]: as far as I can tell up until (and including data) is just the output of `_instrument_data_directory`, so we don't we use that?

File: scripts/Muon/GUI/Common/muon_file_utils.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
+from __future__ import (absolute_import, division, print_function)
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/muon_file_utils.py]: I think we should put this in a new utils folder within Common

File: scripts/Muon/GUI/Common/run_string_utils.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
+from __future__ import (absolute_import, division, print_function)
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/run_string_utils.py]: also in utils folder

File: scripts/Muon/GUI/Common/thread_model.py [hunk 7]
```diff
@@ -1,23 +1,72 @@
 from __future__ import (absolute_import, division, print_function)
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/thread_model.py]: remove this file, as it is being covered in PR #23767

File: scripts/test/Muon/thread_model_test.py [hunk 8]
```diff
@@ -63,17 +63,10 @@ class Runner:
         The finished signal of a QThread is connected to the finished method below"""
```
Comment (AnthonyLim23) [inline scripts/test/Muon/thread_model_test.py]: same as above

File: scripts/test/Muon/utilities/run_string_utils_test.py [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+        for i in range(len(run_lists)):
+            self.assertEqual(utils.run_string_to_list(run_strings[i]), run_lists[i])
+
+    def test_run_string_to_list_orders_non_consecutive_runs(self):
+        run_string = "1-3,48-50"
+        run_list = [1, 2, 3, 48, 49, 50]
+        self.assertEqual(utils.run_string_to_list(run_string), run_list)
+
+
+class RunStringUtilsIncrementAndDecrementRunLists(unittest.TestCase):
```
Comment (AnthonyLim23) [inline scripts/test/Muon/utilities/run_string_utils_test.py]: I think the file is getting a bit long, could we split these into two?

File: scripts/test/Muon/utilities/thread_model_test.py [hunk 10]
```diff
@@ -0,0 +1,179 @@
+import unittest
```
Comment (AnthonyLim23) [inline scripts/test/Muon/utilities/thread_model_test.py]: covered in another PR, so we should remove it here

File: scripts/Muon/GUI/Common/muon_file_utils.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+    return "{0:08d}".format(run)
+
+
+def _instrument_data_directory(instrument):
+    """The directory which stores the data for a particular instrument"""
+    return FILE_SEP + FILE_SEP + get_instrument_directory(instrument) + FILE_SEP + "data"
+
+
+def file_path_for_instrument_and_run(instrument, run):
+    """Returns the path to the data file for a given instrument/run"""
```
[the remark this answers is not in the document]
Comment (AnthonyLim23) [reply scripts/Muon/GUI/Common/muon_file_utils.py]: I am pro simpler

File: scripts/Muon/GUI/Common/run_string_utils.py [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+    """
+    run_list = _remove_duplicates_from_list(run_list)
+    run_list = [i for i in run_list if i >= 0]
+    run_list.sort()
+    if len(run_list) > max_run_list_size:
+        raise IndexError("Too many runs (" + str(len(run_list)) + ") must be <" + str(max_run_list_size))
+
+    range_list = []
+    # use groupby to group run_list into sublists of sequential integers
+    for _, grouped_list in groupby(enumerate(run_list), key=lambda_tuple_unpacking(lambda i, x: i - x)):
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/run_string_utils.py]: can we have a comment explaining this some more

File: scripts/Muon/GUI/Common/muon_file_utils.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Mantid-Matthew) [reply scripts/Muon/GUI/Common/muon_file_utils.py]: Upon discussion it was decided to leave this as is for now.

**Comment by martyngigg** [comment]
Do we think we are nearly ready with this one? I think it's holding up a few of the reviewed PRs from being merged.

**Comment by AnthonyLim23** [comment]
the build failures look to be unrelated

**Comment by AnthonyLim23** [comment]
I would like to give it another look over once the other PRs have been merged

**Comment by martyngigg** [comment]
We've got some conflicts here now that #23645 has been merged.

File: scripts/test/Muon/utilities/muon_load_data_test.py [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+        self.assertEqual(data.contains_n(run=1234), 1)
+        self.assertEqual(data.contains_n(workspace=[1]), 1)
+        self.assertEqual(data.contains_n(filename="C:\\dir1\\file1.nxs"), 1)
+
+    def test_that_clear_empties_the_data_of_all_entries(self):
+        data = MuonLoadData()
+        data.add_data(run=1234, workspace=[1], filename="C:\\dir1\\file1.nxs")
+        data.add_data(run=1235, workspace=[2], filename="C:\\dir1\\file2.nxs")
+        data.add_data(run=1236, workspace=[3], filename="C:\\dir1\\file3.nxs")
+
```
Comment (AnthonyLim23) [inline scripts/test/Muon/utilities/muon_load_data_test.py]: can we check its not empty before clearing it?

File: scripts/test/Muon/utilities/muon_load_data_test.py [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+        data.clear()
+
+        self.assert_empty(data)
+
+    def test_that_adding_then_removing_single_item_leaves_data_empty(self):
+        data = MuonLoadData()
+
+        data.add_data(run=1234, workspace=[1], filename="C:\\dir1\\file1.nxs")
+        data.remove_data(run=1234)
```
Comment (AnthonyLim23) [inline scripts/test/Muon/utilities/muon_load_data_test.py]: same as above

File: scripts/test/Muon/utilities/muon_load_data_test.py [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+        self.assert_empty(data)
+
+    def test_that_remove_method_can_remove_several_entries_at_once(self):
+        data = MuonLoadData()
+
+        data.add_data(run=1234, workspace=[1], filename="C:\\dir1\\file1.nxs")
+        data.add_data(run=1234, workspace=[2], filename="C:\\dir1\\file2.nxs")
+        data.add_data(run=1234, workspace=[3], filename="C:\\dir1\\file3.nxs")
+
+        data.remove_data(run=1234)
```
Comment (AnthonyLim23) [inline scripts/test/Muon/utilities/muon_load_data_test.py]: same a above

File: scripts/test/Muon/utilities/muon_load_data_test.py [hunk 16]
```diff
@@ -0,0 +1,229 @@
...
+
+        self.assert_empty(data)
+
+    def test_that_remove_method_applies_OR_to_multiple_keyword_arguments(self):
+        data = MuonLoadData()
+
+        data.add_data(run=1234, workspace=[1], filename="C:\\dir1\\file1.nxs")
+        data.add_data(run=1235, workspace=[2], filename="C:\\dir1\\file2.nxs")
+        data.add_data(run=1236, workspace=[3], filename="C:\\dir1\\file3.nxs")
+
```
Comment (AnthonyLim23) [inline scripts/test/Muon/utilities/muon_load_data_test.py]: same as above

File: scripts/test/Muon/utilities/muon_load_data_test.py [hunk 17]
```diff
@@ -0,0 +1,229 @@
...
+        data.remove_data(run=1234, workspace=[2], filename="C:\\dir1\\file3.nxs")
+
+        self.assert_empty(data)
+
+    def test_that_contains_is_true_when_data_contains_entry_which_matches_to_a_single_keyword_and_false_otherwise(self):
+        data = MuonLoadData()
+
+        data.add_data(run=1234, workspace=[1], filename="C:\\dir1\\file1.nxs")
+        data.add_data(run=1235, workspace=[2], filename="C:\\dir1\\file2.nxs")
+        data.add_data(run=1234, workspace=[3], filename="C:\\dir1\\file3.nxs")
```
Comment (AnthonyLim23) [inline scripts/test/Muon/utilities/muon_load_data_test.py]: the 3 above lines seem to be repeated a lot. Could we refactor them out to reduce repeated code?

File: scripts/test/Muon/utilities/run_string_utils_conversion_test.py [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+    def test_run_string_to_list_removes_duplicates(self):
+        run_strings = ["5,5,5,5,4,3,2,1,1", "1,2,4,5,1-3,1-5,1-4"]
+        for run_string in run_strings:
+            run_list = utils.run_string_to_list(run_string)
+            self.assertEqual(run_list, [1, 2, 3, 4, 5])
+
+    def test_run_string_to_list_throws_for_incorrectly_placed_range_separator(self):
+        run_strings = ["-1,2,3", "1,2,3-"]
+        for run_string in run_strings:
+            with self.assertRaises(IndexError) as context:
```
Comment (AnthonyLim23) [inline scripts/test/Muon/utilities/run_string_utils_conversion_test.py]: is context the best variable name?

**Review by AnthonyLim23 — APPROVED** [review]

**Pull request merged**